MatrixSDK/Crypto/MXCrypto.m

```diff
@@ -1964,27 +1976,118 @@ - (MXHTTPOperation*)ensureOlmSessionsForDevices:(NSDictionary<NSString* /* userI
 
     if (devicesWithoutSession.count == 0)
     {
+        NSLog(@"[MXCrypto] ensureOlmSessionsForDevices: Have already sessions for all");
         if (success)
         {
             success(results);
         }
         return nil;
     }
 
+    
     NSString *oneTimeKeyAlgorithm = kMXKeySignedCurve25519Type;
-
-    // Prepare the request for claiming one-time keys
+    
+    // Devices for which we will make a /claim request
     MXUsersDevicesMap<NSString*> *usersDevicesToClaim = [[MXUsersDevicesMap<NSString*> alloc] init];
+    // The same but devices are listed by their identity key
+    NSMutableArray<NSString*> *devicesToClaim = [NSMutableArray array];
+    
+    // Devices (by their identity key) that are waiting for a response to /claim request
+    // That can be devices for which we are going to make a /claim request OR devices that
+    // already have a pending requests.
+    // Once we have emptied this array, we can call the success or the failure block. The
+    // operation is complete.
+    NSMutableArray<NSString*> *devicesInProgress = [NSMutableArray array];
+    
+    // Prepare the request for claiming one-time keys
     for (MXDeviceInfo *device in devicesWithoutSession)
     {
-        [usersDevicesToClaim setObject:oneTimeKeyAlgorithm forUser:device.userId andDevice:device.deviceId];
+        NSString *deviceIdentityKey = device.identityKey;
+        
+        // Claim only if a request is not yet pending
+        if (![ensureOlmSessionsInProgress containsObject:deviceIdentityKey])
+        {
+            [usersDevicesToClaim setObject:oneTimeKeyAlgorithm forUser:device.userId andDevice:device.deviceId];
+            [devicesToClaim addObject:deviceIdentityKey];
+            
+            [ensureOlmSessionsInProgress addObject:deviceIdentityKey];
+        }
+        
+        [devicesInProgress addObject:deviceIdentityKey];
```

Comment:
I would move this line `[devicesInProgress addObject:deviceIdentityKey];` at the beginning of the block instead of at the end (to highlight that I don't need to know whether a request is in progress or not)